### PR TITLE
Add GetSupportedFirmwareUpdateMethods to redfish_facts/Update commands

### LINKED ISSUE #54234: redfish_facts - Update - GetSupportedFirmwareUpdateMethods
##### SUMMARY

This feature would implement a GetSupportedFirmwareUpdateMethods (or something less verbose) command for the Update category of redfish_facts, and would retrieve a list of the methods by which the system supports performing firmware updates.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

redfish_facts

##### ADDITIONAL INFORMATION

```yaml

```

### PR DESCRIPTION
**Pull request by xmadsen**
##### SUMMARY

This pull request implements GetSupportedFirmwareUpdateMethods, which returns as a list the supported firmware update methods as shown in the UpdateService schema. 

Fixes #54234 

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

redfish_facts
redfish_utils

##### ADDITIONAL INFORMATION

```yaml
- name: Test Redfish modules
  hosts: localhost
  gather_facts: false
  vars:
    baseuri: 10.243.5.31
    user: USERID
    password: PASSW0RD

  tasks:

    - name: Get all information available in Update category
      redfish_facts:
        category: Update
        command: GetSupportedFirmwareUpdateMethods
        baseuri: "{{ baseuri }}"
        username: "{{ user }}"
        password: "{{ password }}"
```

```paste below
 ansible-playbook 2.8.0.dev0
  config file = None
  configured module search path = [u'/Users/xandermadsen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/xandermadsen/ansible/lib/ansible
  executable location = /Users/xandermadsen/ansible/bin/ansible-playbook
  python version = 2.7.15 (default, Feb 12 2019, 11:00:12) [GCC 4.2.1 Compatible Apple LLVM 10.0.0 (clang-1000.11.45.5)]
No config file found; using defaults
host_list declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
script declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
auto declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
yaml declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
ini declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
Skipping due to inventory source not existing or not being readable by the current user
toml declined parsing /etc/ansible/hosts as it did not pass it's verify_file() method
 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAYBOOK: test_redfish_facts.yml **************************************************************************************************************************************************************************
1 plays in ../test_redfish_facts.yml

PLAY [Test Redfish modules] *******************************************************************************************************************************************************************************
META: ran handlers

TASK [Get all information available in Update category] ***************************************************************************************************************************************************
task path: /Users/xandermadsen/test_redfish_facts.yml:11
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: xandermadsen
<127.0.0.1> EXEC /bin/sh -c 'echo ~xandermadsen && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697 `" && echo ansible-tmp-1553292386.15-246599246160697="` echo /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697 `" ) && sleep 0'
Using module file /Users/xandermadsen/ansible/lib/ansible/modules/remote_management/redfish/redfish_facts.py
<127.0.0.1> PUT /Users/xandermadsen/.ansible/tmp/ansible-local-419185AEjdJ/tmpDvBITr TO /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697/AnsiballZ_redfish_facts.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697/ /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697/AnsiballZ_redfish_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '/usr/local/opt/python@2/bin/python2.7 /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697/AnsiballZ_redfish_facts.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/xandermadsen/.ansible/tmp/ansible-tmp-1553292386.15-246599246160697/ > /dev/null 2>&1 && sleep 0'
ok: [localhost] => (item=GetFirmwareInventory) => {
    "ansible_facts": {
        "redfish_facts": {
            "firmware_update_methods": {
                "entries": [
                    "TFTP", 
                    "SFTP"
                ], 
                "ret": true
            }
        }
    }, 
    "changed": false, 
    "failed_when_result": false, 
    "invocation": {
        "module_args": {
            "baseuri": "10.243.5.31", 
            "category": [
                "Update"
            ], 
            "command": [
                "GetSupportedFirmwareUpdateMethods"
            ], 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "username": "USERID"
        }
    }, 
    "item": "GetFirmwareInventory"
}
META: ran handlers
META: ran handlers

PLAY RECAP ************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

```

### CODE REVIEW

**Comment by xmadsen** [comment]
@samerhaj

**Comment by samerhaj** [comment]
Some feedback:

* We should return whether the service supports updating via SimpleUpdate Action, and/or HttpPushUri, or other methods in the future 

* The returned protocols (TFTP/SFTP/etc...) are for SimpleUpdate

* Maybe rename to GetFirmwareUpdateCapabilities ?

File: lib/ansible/module_utils/redfish_utils.py
Comment (billdodd) [inline lib/ansible/module_utils/redfish_utils.py]: Rather than hardcode "/redfish/v1/UpdateService", better to read the value of the "@odata.id" property from the "UpdateService" property.

**Review by billdodd — COMMENTED** [review]
Agree with @samerhaj's comments, including the command name suggestion of `GetFirmwareUpdateCapabilities`.

Also, see one other comment inline about the UpdateService URI.

File: lib/ansible/module_utils/redfish_utils.py
Comment (billdodd) [inline lib/ansible/module_utils/redfish_utils.py]: Neither the 'title' nor the 'TransferProtocol@Redfish.AllowableValues' properties are required, so these are likely to cause KeyError exceptions. Suggest changing the `for` loop to something like this:

```
        for key in actions.keys():
            action = actions.get(key)
            if 'title' in action:
                title = action['title']
            else:
                title = key
            result['entries'][title] = action.get(
                'TransferProtocol@Redfish.AllowableValues',
                ["key TransferProtocol@Redfish.AllowableValues not found"]
            )
```

**Review by billdodd — COMMENTED** [review]
Thanks for the update @xmadsen.  A couple of suggestions:

1. In general, when adding a new command, good to add an example in the EXAMPLES section of the module.

2. See inline comment below for suggestion to avoid KeyError exceptions.

**Review by billdodd — APPROVED** [review]
Thanks! Looks good.

shipit

File: lib/ansible/module_utils/redfish_utils.py
Comment (jose-delarosa) [inline lib/ansible/module_utils/redfish_utils.py]: Is it possible to not use "/redfish/v1" here but instead pass it from `redfish_facts` like we do in most other methods? (Nice to be consistent)

Comment (jose-delarosa) [inline lib/ansible/module_utils/redfish_utils.py]: Is it possible to not use `"/redfish/v1"` here but instead pass it from the module like we do with most other methods?

**Review by jose-delarosa — CHANGES_REQUESTED** [review]
See comments.

On Dell servers the key `TransferProtocol@Redfish.AllowableValues` is not found, but it clearly detects that is not found.

File: lib/ansible/module_utils/redfish_utils.py
Comment (billdodd) [reply lib/ansible/module_utils/redfish_utils.py]: Good catch on the `"/redfish/v1"`. I agree.

**Comment by xmadsen** [comment]
@billdodd @jose-delarosa this should be good to check

**Review by billdodd — APPROVED** [review]
Looks good.

shipit

**Comment by billdodd** [comment]
@xmadsen - may need to add `ready_for_review` comment to clear the `needs_revision` label.

**Comment by xmadsen** [comment]
ready_for_review

**Comment by jose-delarosa** [comment]
I am not able to test this patch in my dev environment. I've tried 2 ways:

1) As described [here](https://help.github.com/en/articles/checking-out-pull-requests-locally).
2) Download [patch](https://patch-diff.githubusercontent.com/raw/ansible/ansible/pull/54268.patch) and apply with either `git am` or `git apply`.

Open to suggestions.

**Comment by billdodd** [comment]
@jose-delarosa I just downloaded the 2 changed files into my devel branch sandbox and was able to run successfully.

**Comment by jose-delarosa** [comment]
@billdodd I tried again, no luck. Not sure what is going on, I've been able to test every PR so far. I rebase my dev branch with upstream and then apply the patch in a separate branch, no luck. Will try again.

If @mraineri or @tomasg2012 can test and approve, I'm OK with that.

**Comment by billdodd** [comment]
> @billdodd I tried again, no luck. Not sure what is going on, I've been able to test every PR so far. I rebase my dev branch with upstream and then apply the patch in a separate branch, no luck. Will try again.

What kind of error(s) are you getting? I assume it must be an issue related to a recent merge. But there hasn't been very many, so it seems like it shouldn't be too hard to work around. If I saw the error, I might be able to give some input.

> If @mraineri or @tomasg2012 can test and approve, I'm OK with that.

Absolutely. That would be good too.

**Comment by jose-delarosa** [comment]
@xmadsen @billdodd  When I apply this patch to my devel branch and try to test it, I get HTTP Error 406, and sure enough my patch https://github.com/ansible/ansible/pull/55193 for adding those extra HEADERS is not there. Looks to me like you originally submitted this PR before https://github.com/ansible/ansible/pull/55193, but not sure why this would cause a problem since these are different blocks of code. Could be a GitHub bug.

In any regard, I am not able to test as-is. Unless you want to rebase your repo to inculde the HEADERS PR?

**Comment by billdodd** [comment]
@jose-delarosa Could you test on a 14G system with the previous firmware or a 13G system?

**Comment by jose-delarosa** [comment]
Well best I can do is 13G, and at least the key-not-found message is captured correctly:

```
{
    "ansible_facts": {
        "redfish_facts": {
            "firmware_update_capabilities": {
                "entries": {
                    "#UpdateService.SimpleUpdate": [
                        "Key TransferProtocol@Redfish.AllowableValues not found"
                    ]
                },
                "ret": true
            }
        }
    },
    "changed": false,
    "failed": false
}

```
I'm OK with it.

shipit

**Comment by jose-delarosa** [comment]
@billdodd @xmadsen There are about 4 unmerged PRs with 2 shipt its, wonder if we need to reach out to repo maintainers?

**Comment by billdodd** [comment]
> @billdodd @xmadsen There are about 4 unmerged PRs with 2 shipt its, wonder if we need to reach out to repo maintainers?

I'll ping the ansible-devel channel

**Comment by xmadsen** [comment]
Thanks!

**Comment by xmadsen** [comment]
@gundalow this is ready to be merged

**Comment by gundalow** [comment]
Merged into devel for release in Ansible 2.9
It's great to see such a community of reviewers around the Remote Management PRs

Generally speaking, once a PR gets `label/shipit` it will be merged in a few days (to a week)

**Pull request merged**
